File: src/handler.ts
import { APIGatewayProxyEventV2, APIGatewayProxyResultV2 } from 'aws-lambda';
import { scrapePost } from './scraper';
import { extractEventData } from './gen-ai';
import { EventData } from './types';

export const importPost = async (event: APIGatewayProxyEventV2): Promise<APIGatewayProxyResultV2> => {
  try {
    const url = event.queryStringParameters?.url;

    if (!url) {
      return {
        statusCode: 400,
        body: JSON.stringify({ error: 'A valid IG URL is required' }),
      };
    }

    // scraper
    const { imageBase64, caption } = await scrapePost(url);

    // gen-ai
    const eventData: EventData = await extractEventData(imageBase64, caption);

    return {
      statusCode: 200,
      body: JSON.stringify(eventData),
    };
  } catch (error) {
    let errorMessage;

    if (error instanceof Error) {
      errorMessage = error.message;
    } else {
      errorMessage = 'An unexpected error occurred';
    }

    return {
      statusCode: 500,
      body: JSON.stringify({ error: errorMessage }),
    }
  }
};

File: src/handler.py
import json
import time

from src.gen_ai import extract_event_data
from src.scraper import scrape_post


def import_post(event, context):
    try:
        params = event.get("queryStringParameters") or {}
        url = params.get("url")

        if not url:
            return {
                "statusCode": 400,
                "body": json.dumps({"error": "Missing required query parameter: url"}),
            }

        timing = {}
        t_start = time.perf_counter()

        t = time.perf_counter()
        scraped = scrape_post(url)
        timing["scrape_ms"] = round((time.perf_counter() - t) * 1000)

        t = time.perf_counter()
        event_data = extract_event_data(scraped["imageUrl"], scraped["caption"])
        timing["gemini_ms"] = round((time.perf_counter() - t) * 1000)

        timing["total_ms"] = round((time.perf_counter() - t_start) * 1000)

        return {
            "statusCode": 200,
            "body": json.dumps({**event_data, "_timing": timing}),
        }

    except Exception as e:
        return {
            "statusCode": 500,
            "body": json.dumps({"error": str(e)}),
        }
